js/k6/exec: Set VU Tags

codebien opened the pull request:
Extended the `k6/execution` API to allow the get or set of a Tag in the VU's state.

This part of the [Stage tagging work](https://github.com/grafana/k6/issues/796), but it can be considered as an independent feature.

```js
import http from 'k6/http';
import exec from 'k6/execution';

export const options = {
  stages: [
    { duration: '1m', target: 10 },
    { duration: '1m', target: 20 },
    { duration: '1m', target: 0 },
  ],
};

export default function () {
  exec.vu.tags['stage'] = 'stage#2'
  console.log(exec.vu.tags['stage']) // returns stage#2
}
```

<!--
  (ﾉ◕ヮ◕)ﾉ*:・ﾟ✧
  
  Thank you for your interest in contributing to the k6 project!
  
  Before you get started, we'd kindly like to ask you to read our:
    - Contribution guidelines at https://github.com/grafana/k6/blob/master/CONTRIBUTING.md
    - Code of Conduct at https://github.com/grafana/k6/blob/master/CODE_OF_CONDUCT.md
    
  Out of respect for your time, please start a discussion regarding any bigger contributions either
  in a GitHub Issue, in the community forums or in the #contributors channel of the k6 slack before you
  get started on the implementation.
  
  If you've already done all of that, you're more than welcome to proceed with your pull request.
  Thank you again for your contribution! 🙏🏼
  
  
-->

MStoykov commented on js/modules/k6/execution/execution_test.go:
```diff
@@ -0,0 +1,79 @@
+	// get
+	tag, err = rt.RunString(`exec.vu.tags["custom-tag"]`)
+	require.NoError(t, err)
+	assert.Equal(t, "mytag", tag.String())
```
Sorry, for being late to the party, but I missed this last week. 

I kind of feel like just going to string with w/e is given is the better approach. After all we are in javascript and IME this is what is happening everywhere in js land all the time. 

As you've noticed it's very likely that a user will like to set a number and having them get an exception(or silently skipping the set) is likely not the UX that is better. Arguably we can just support numbers and string, but I feel like just calling `String()` on the Value we get will be fine. We are doing the same (or just taking a `string` which under the hood does the same) in a lot of places and rarely had anything but the above problem of `[Object object]`. 

Also arguably this can be discussed in an issue and pulling the people who either write scripts on daily bases and/or people who write js to give their opinion ;)

codebien replied:
>Arguably we can just support numbers and string

I don't think we can do that, we are storing in a `map[string]string` so it would be impossible for the `Get` case to revert in the right type. Except, if we would move to `map[string]interface{}`.

@MStoykov @yorugac I checked the current tags system and apparently the following example is already supported:

```js
export default function () {
  http.get('http://httpbin.org', {tags: {my_tag: "I'm a tag"}});
  http.get('http://test.k6.io', {tags: {my_tag: 4}}); // implicit conversion here
}
```
so except if we would break the API or have an inconsistent behaviour across them, I think we should go with the original design having freedom and an implicit conversion to String.

MStoykov replied:
> I don't think we can do that, we are storing in a map[string]string so it would be impossible for the Get case to revert in the right type. Except, if we would move to map[string]interface{} but I don't think we need that complexity here.

I meant only from the perspective of setting it so that `exec.iteration.tags["something"] = 5` is fine, but looking at goja's documentation it will be kind of hard to make it and arguably, again, we already convert to string in a lot of places (including the relevant place for `tags` as you have noticed) so IMO we should just be calling `String()` (or `ToString()` not certain which one it was) and go with the flow of js :shrug:

yorugac replied:
For my understanding, what about `val.ExportType()`? Can't that be used to get the `val is Object` condition? I've looked at `ExportType` only briefly but it seems it should return `map[string]interface{}` only for JS objects...
Btw, for arrays it should return `[]interface{}` apparently:
```golang
var (
	reflectTypeInt    = reflect.TypeOf(int64(0))
	reflectTypeBool   = reflect.TypeOf(false)
	reflectTypeNil    = reflect.TypeOf(nil)
	reflectTypeFloat  = reflect.TypeOf(float64(0))
	reflectTypeMap    = reflect.TypeOf(map[string]interface{}{})
	reflectTypeArray  = reflect.TypeOf([]interface{}{})
	reflectTypeString = reflect.TypeOf("")
)
```

codebien replied:
>Can't that be used to get the val is Object condition?

Ok, but in the hypothesis, we could get a response for this, what is the end goal? Why should we treat a JS Object in a different way from a Boolean or a Number?

yorugac replied:
To help avoid garbage values in the tags. I.e. booleans and numbers will be implicitly converted to something sensible and recognizable, even if not always precise, but `[object Object]` as tag value is ~ 100% a mistake and not what user wants to see in the tags. 
I think if Objects are allowed here, we might end up having to add additional filtering on tags in other places, like output extensions, or in whatever implementation for tags appears during re-factoring.

Granted, the fact that this is already allowed in the normal tags is a point... :thinking: Though it leads to conclusion that it's one of the things to improve from UX perspective there too :slightly_smiling_face: But I guess that'd be out of scope of this PR so I can agree that this solution is fine here, for now at least.

na-- replied:
FWIW, we might want to do something similar to what @MStoykov did for adding values to custom metrics recently: https://github.com/grafana/k6/pull/1876#issuecomment-946753448 In that case, we expect only numbers. Adding anything that wasn't a number introduced problems, so now doing it will be discarded with a warning by default, but throw an exception if the `throw` option is enabled. 

In this case, we can probably be a bit more liberal and convert numbers and bools to string automatically. But @yorugac is completely right that having an object or an array here is ~100% a mistake we probably shouldn't silently accept. I am fine with making an issue about this and fixing it in a separate PR though. If we document that users are supposed to use only string/int/bool values, it won't be a breaking change.

MStoykov replied:
I kind of ... disagree here as there I was moving to numbers/booleans, but here we move to `string` and there is [Object.prototype.toString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toString), so it's technically possible to set a tag value that is an object or array or whatever and have the `toString` method of it be set to appropriate (either on the prototype of the thing or only on the particular instance) and that will be expected to lead to the method being called and used IMO ;).

But let's discuss this in an issue and again, maybe pulling some js people and people who write scripts to give their opinion

na-- replied:
:thinking:  true, you can define `toString` on your own objects, but as far as I know, there isn't a way to detect if the user has implemented it or if it will just default to the base `[object Object]`? So, overall, I think it's much safer to go with the explicit approach here, instead of silently converting everything to string. Users can easily call `.toString()` on their objects manually, if that's what they want, and the majority of the time trying to pass objects as strings is probably going to be a mistake. And, if someone complains, we can more easily relax the restriction than implement a new one... 

re: discussing this in an issue, what's the problem to discuss it here? This PR implement just a part of an issue, and this specific problem is is an implementation detail of this PR, so the PR is as good of a place to discuss it as any...

codebien replied:
@na-- `v0.34.1` is accepting something like:
```js
export default function () {
  http.get('http://httpbin.org', {tags: {my_tag: {name: "bob", age: 25}}});
}

// http_req_duration.my_tag
// [object Object]
```
Doing Discard+Warning in this PR will generate something not consistent with the previous case. Should we change both then?

na-- replied:
:disappointed: :thinking: 

For this PR, I  think you should just make `exec.vu.tags` work like this, everything else is out of scope. This is a new feature and it won't be a breaking change to have more strict validation when it's launched.

For the `tags` parameter in HTTP requests, custom metrics, etc., please make a new issue to discuss how, if and when we should address it. We might decide to follow @MStoykov's suggestion and just do nothing, continue silently casting tag values to strings. Then we can also relax the validation here, if we decide to. Or we might decide to make the minor breaking change and introduce the stricter validation there as well :man_shrugging: But it definitely deserves more discussion and input from JS developers, and is definitely out of scope for this PR.

codebien replied:
ok, for summary:
* we accept `string`, `bool`, `number`
* if another kind is provided then discard and warn about it
* if `--throw` then raise an exception

na-- replied:
:+1:, + create a new issue to discuss this further, potentially adressing the same problem in other places or relaxing this validation here

na-- replied:
New issue is https://github.com/grafana/k6/issues/2218, resolving this thread

codebien commented:
Removed the latest fixup with the type limit instead added documentation about conversion in the `Set` method's documentation.

MStoykov reviewed (approved):
LGTM

na-- reviewed (commented):
Mostly LGTM and I'm fine with merging it if you make a v0.36.0 issue about https://github.com/grafana/k6/pull/2172/files#r740787612

Though fixing that should probably be easy, so it might be a bit better if you do it in this PR :man_shrugging:

codebien commented:
> Though fixing that should probably be easy, so it might be a bit better if you do it in this PR :man_shrugging:

@na-- I prefer to address it now, I will try it today :+1:

codebien commented:
Pushed changes as defined in https://github.com/grafana/k6/pull/2172#discussion_r740996518

na-- commented on js/modules/k6/execution/execution.go:
```diff
@@ -208,3 +219,81 @@ func newInfoObj(rt *goja.Runtime, props map[string]func() interface{}) (*goja.Ob
+			common.Throw(o.Runtime, err)
+			return false
+		}
+		o.State.Logger.Warn("the Set operation has been discarded because" + err.Error())
```
```suggestion
		o.State.Logger.Warnf("the execution.vu.tags.Set('%s') operation has been discarded because %s", key, err.Error())
```

na-- commented on js/modules/k6/execution/execution.go:
```diff
@@ -208,3 +219,81 @@ func newInfoObj(rt *goja.Runtime, props map[string]func() interface{}) (*goja.Ob
+func (o *tagsDynamicObject) Set(key string, val goja.Value) bool {
+	switch val.ExportType().Kind() { //nolint:exhaustive
+	case
+		reflect.String,
+		reflect.Bool,
+		reflect.Int,
+		reflect.Int8,
+		reflect.Int16,
+		reflect.Int32,
+		reflect.Int64,
+		reflect.Uint,
+		reflect.Uint8,
+		reflect.Uint16,
+		reflect.Uint32,
+		reflect.Uint64,
+		reflect.Float32,
+		reflect.Float64,
+		reflect.Complex64,
+		reflect.Complex128:
```
```suggestion
		reflect.String,
		reflect.Bool,
		reflect.Int64,
		reflect.Float64,
```
I am [pretty sure goja doesn't use any of the others](https://github.com/dop251/goja/blob/dc8c55024d06009a0cba080cd738192d368735de/value.go#L85-L87), but cc @MStoykov WDYT?

codebien replied:
I already applied the change because the linked goja's code makes the suggestion reasonable. I will wait for the @MStoykov's opinion before rebasing.

MStoykov replied:
Given [this list](https://github.com/dop251/goja/blob/dc8c55024d06009a0cba080cd738192d368735de/value.go#L44) of predefined reflect types that then get returned by `ExportType` it seems like this should be enough. 

I had some suspicion that maybe if it goes down some rabbit hole of reflecting go objects it might return the original int type but testing with `struct{X int8}` still gives you `int64` from my quick testing, so I guess that isn't a problem as well

MStoykov commented on js/modules/k6/execution/execution.go:
```diff
@@ -208,3 +219,69 @@ func newInfoObj(rt *goja.Runtime, props map[string]func() interface{}) (*goja.Ob
+// In any other case, if the Throw option is set then an error is raised
+// otherwise just a Warning is written.
+func (o *tagsDynamicObject) Set(key string, val goja.Value) bool {
+	switch val.ExportType().Kind() { //nolint:exhaustive
```
Doesn't the `default` case make this linter compliment?

MStoykov replied:
https://github.com/nishanths/exhaustive/issues/1

codebien commented on js/modules/k6/execution/execution_test.go:
```diff
@@ -0,0 +1,169 @@
+func TestVUTags(t *testing.T) {
+	t.Parallel()
+
+	setupExecModule := func(t *testing.T) (*ModuleInstance, *goja.Runtime, *testutils.SimpleLogrusHook) { // nolint:unparam
```
@MStoykov regarding linter, this also is not totally clear to me, maybe it's spotting a bug that I'm not seeing

MStoykov replied:
You are not using `t` I think

MStoykov replied:
from my local experiments ... it seems like `unparam` can't find that you actually use the hook(the last result) in one of the cases and that is what is being.

If you add it to first subcase and assert that it's empty it fixed itself :shrug: 

Also maybe (not necessary in this PR you can move this to the outside in a function and not actually have sub tests with but just a bunch of single tests as this just adds indentation currently. 

I also have started to make these functions to return a struct with the things so that you don't have to constantly add 1 more returned thing whenever we add one more thing we need.

MStoykov commented on js/modules/k6/execution/execution_test.go:
```diff
@@ -0,0 +1,169 @@
+			require.NoError(t, err)
+			assert.Equal(t, "true", val.String())
+
+			_, err = rt.RunString(`exec.vu.tags["vu"] = 101`)
+			require.NoError(t, err)
+			val, err = rt.RunString(`exec.vu.tags["vu"]`)
+			require.NoError(t, err)
+			assert.Equal(t, "101", val.String())
```
maybe a float test as well?

MStoykov reviewed (dismissed):
LGTM :clap: 

I would just like to point out that for some system tags setting them through this will have effects.

Like for example setting `name` will work for HTTP requests, but setting `url` (and having it still enabled in system tags) will not, and probably most others in such a situation. But for example `vu` or `scenario` will always be overwritten.

This definitely needs some documentation and likely at least a discussion (in an issue) on what needs to be done if anything

MStoykov commented on js/modules/k6/execution/execution.go:
```diff
@@ -208,3 +219,69 @@ func newInfoObj(rt *goja.Runtime, props map[string]func() interface{}) (*goja.Ob
+		reflect.String,
+		reflect.Bool,
+		reflect.Int64,
+		reflect.Float64,
```
```suggestion
		reflect.Float64:
```

codebien commented:
* rebased
* refactor of the helper
* added a float case

>This definitely needs some documentation and likely at least a discussion (in an issue) on what needs to be done if anything
I will add it to the documentation, @MStoykov can you open an issue for it, please? (I'm not sure to know all the possible cases so I will be for sure not exhaustive in describing the issue, and probably you already have a better vision about possible implementations).

yorugac reviewed (approved):
This version looks much more neat. Nice work :smile:
